pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -39,284 +108,413 @@ func Add(mgr manager.Manager) error {
-	//The length of time to wait before giving up, zero means infinite
-	r.drainer.Timeout = time.Minute
+					if err := r.client.Update(context.TODO(), lease); err != nil {
+						r.reqLogger.Error("Failed to update the lease %v", err)
 
-	// TODO - consider pod selectors (only for VMIs + others ?)
-	//Label selector to filter pods on the node
-	//r.drainer.PodSelector = "kubevirt.io=virt-launcher"
+						statusAnnoData.status = NodeStateFailedToExtendLease
+						_, err = r.setStatusAnnotation(node, statusAnnoData)
 
-	cs, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		return err
-	}
-	r.drainer.Client = cs
-	r.drainer.DryRun = false
+						return reconcile.Result{}, err
+					}
+				}
+			}
+			fallthrough
 
-	return nil
-}
+		case LeaseStatusActiveOwnedByMe:
+			{
 
-var _ reconcile.Reconciler = &ReconcileNodeMaintenance{}
+				if statusAnnoData.status != NodeStateCordon && statusAnnoData.status != NodeStateEvicting {
+					err := fmt.Errorf("Illegal state. node %s state %s", request.Name, statusAnnoData.status)
+					r.reqLogger.Errorf("error %v", err)
+					return reconcile.Result{}, err
+				}
 
-type ReconcileHandler interface {
-	StartPodInformer(node *corev1.Node, stop <-chan struct{}) error
-}
+				r.reqLogger.Infof("holding lease. Maintenance mode ongoing. node: %s\n", request.Name)
+			}
 
-// ReconcileNodeMaintenance reconciles a NodeMaintenance object
-type ReconcileNodeMaintenance struct {
-	// This client, initialized using mgr.Client() above, is a split client
-	// that reads objects from the cache and writes to the apiserver
-	client      client.Client
-	scheme      *runtime.Scheme
-	drainer     *drain.Helper
-	podInformer cache.SharedInformer
-}
+		case LeaseStatusActiveOwnedByDifferentNode:
+			{
+				if statusAnnoData.status == NodeStateInit {
+					r.reqLogger.Infof("wait to acquire lease. requeue. node: %s\n", request.Name)
+					r.setStatusAnnotation(node, statusAnnoData)
+					return reconcile.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+				}
 
-var Handler ReconcileHandler
+				statusAnnoData.status = NodeStateLeaseNotOwnedByNode
+				r.setStatusAnnotation(node, statusAnnoData)
 
-// Reconcile reads that state of the cluster for a NodeMaintenance object and makes changes based on the state read
-// and what is in the NodeMaintenance.Spec
-// Note:
-// The Controller will requeue the Request to be processed again if the returned error is non-nil or
-// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
-func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	reqLogger := log.WithFields(log.Fields{"Request.Namespace": request.Namespace, "Request.Name": request.Name})
-	reqLogger.Info("Reconciling NodeMaintenance")
+				err := fmt.Errorf("lease got acquired by someone while request was active. node: %s", request.Name)
+				return reconcile.Result{}, err
```

Comment:
Nothing we should do here?

Reply:
two lines earlier I am setting an error state. Please suggest what can be done besides this.  

                        statusAnnoData.status = NodeStateLeaseNotOwnedByNode                                                                                                                  
                        r.setStatusAnnotation(node, statusAnnoData)